### PR TITLE
Optimize sorting tree container lookup

### PR DESCRIPTION
**Pull request by KaroUniform**
## What changed

- Enumerate block entities from already-loaded chunks instead of probing every block position in the configured cube.
- Cache sided item capabilities with NeoForge's `BlockCapabilityCache` while the target block entity is loaded.
- Partition capability caches by `ServerLevel` identity, use weak block-entity keys within each level, and clear the complete per-level table on server-side `LevelEvent.Unload`.
- Update the sorting-particle call to the 26.1 packet signature.
- Add coverage for chunk boundaries, unloaded chunks, removed block entities, successful transfers, transaction rollback, directional cache reuse, cross-level isolation, and selective level cleanup.

## Why

The sorting tree runs once per second. At the default range of 16, the previous implementation checked 35,937 block positions per active core before it could begin sorting. The loaded-chunk lookup does not load or generate chunks and scales with loaded chunks and block entities rather than the volume of the search cube.

The follow-up keeps NeoForge's recommended `BlockCapabilityCache` for repeated queries without restoring the old lifetime and cross-level keying problems. Caches are separated by level, ordinary block/chunk invalidation releases cached handlers, and level unload explicitly removes the remaining per-level object graph.

This is the forward port of #2669 to the current `latest` / 26.1 branch.

## Validation

- Java 25 strict compilation with the project's `-Xlint` flags and `-Werror`: passed for the changed production code and tests against the real Minecraft 26.1.1 / NeoForge 26.1.1.6-beta classpath.
- Standalone lifecycle harness: passed cache reuse, directional separation, identical positions in different levels, selective clear, cache recreation, and position changes.
- Real NeoForge dedicated-server integration test with chests at the same coordinates in the Overworld and Nether: passed reuse, directional separation, level isolation, and selective cleanup.
- 500,000 cached capability lookups completed in 17 ms without increasing cache-table sizes.
- Clean server shutdown observed per-level unload cleanup from 2 tables to 1 and then 0.
- Independent correctness and red-team reviews found no blockers.
- `git diff --check`: passed.

The full upstream `latest` build is not currently a usable signal: the base branch has unrelated unfinished 26.1 port compilation errors, and its Azure pipeline still selects JDK 21 while the Gradle toolchain requires Java 25. The isolated checks above avoid hiding that baseline condition while still compiling and exercising the changed code.

### CODE REVIEW

File: src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java [hunk 1]
```diff
@@ -1,31 +0,0 @@
-package twilightforest.util;
```
Comment (Tamaized) [inline src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java:1]: Why was this deleted?

Comment (KaroUniform) [reply src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java:1]: It was only used by the Sorting Tree (and its dedicated test). This PR no longer needs it because, while iterating loaded block entities, we already have the `BlockEntity` and `BlockState` and can query the capability directly. Leaving it would therefore be dead code.

Its map also lived for the lifetime of the block singleton without eviction, was keyed only by position and direction (not by `ServerLevel`), and the previous cube iterator could supply mutable `BlockPos` instances. Retaining it would preserve the lifetime and cross-level keying issues this change is intended to remove. The test was removed along with the now-unused helper.

Comment (Tamaized) [reply src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java:1]: NeoForge devs strongly recommend to use BlockCapabilityCache whenever possible.

Comment (Tamaized) [reply src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java:1]: Please read: https://docs.neoforged.net/docs/inventories/capabilities#block-capability-caching

Comment (KaroUniform) [reply src/main/java/twilightforest/util/BlockCapabilityDirectionalCache.java:1]: Thanks for pointing this out and for linking the NeoForge guidance. I reworked this in 7c0c4d867 to keep using `BlockCapabilityCache` for each loaded block entity and side.

To avoid the lifetime and cross-level issues in the previous helper, the caches are now partitioned by `ServerLevel`, use weak block-entity keys within each level, and explicitly clear the complete per-level table on server-side `LevelEvent.Unload`.

I added tests for reuse, directional separation, identical positions in different levels, and selective cleanup. I also exercised the exact cache implementation on a NeoForge dedicated server: 500,000 repeated lookups kept the table sizes stable, and shutdown cleanup went from 2 level tables to 1 and then 0.

Could you please take another look and let me know whether this now matches the recommended pattern?

File: src/main/java/twilightforest/events/SortingTreeEvents.java
Comment (a6tumn) [inline src/main/java/twilightforest/events/SortingTreeEvents.java:1]: Not sure if a single event needs its own class, would possibly fit better into *MiscEvents*.

File: src/test/java/twilightforest/util/WorldUtilTests.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		List<BlockEntity> result = WorldUtil.getLoadedBlockEntitiesInRange(level, new BlockPos(-16, 64, -16), 1);
+
+		assertEquals(2, result.size());
+		assertEquals(Set.of(negativeEdge, positiveEdge), new HashSet<>(result));
+		verify(chunkSource).getChunkNow(-2, -2);
+		verify(chunkSource).getChunkNow(-1, -2);
+		verify(chunkSource).getChunkNow(-2, -1);
+		verify(chunkSource).getChunkNow(-1, -1);
+		verifyNoMoreInteractions(chunkSource);
+	}
```
Comment (Sleepy-Horse) [inline src/test/java/twilightforest/util/WorldUtilTests.java:92]: Why do we test it for chunks with negative coordinates? Do you consider it an edge case or have I misunderstood the purpose of the test?

File: src/main/java/twilightforest/util/WorldUtil.java [hunk 3]
```diff
@@ -61,6 +63,35 @@ public static Iterable<BlockPos> getAllAround(BlockPos center, int range) {
...
+			.filter(Objects::nonNull)
+			.flatMap(chunk -> chunk.getBlockEntities().values().stream())
+			.filter(blockEntity -> !blockEntity.isRemoved())
+			.filter(blockEntity -> {
+				BlockPos blockPos = blockEntity.getBlockPos();
+				return blockPos.getX() >= minX && blockPos.getX() <= maxX
+					&& blockPos.getY() >= minY && blockPos.getY() <= maxY
+					&& blockPos.getZ() >= minZ && blockPos.getZ() <= maxZ;
+			})
+			.toList();
```
Comment (Sleepy-Horse) [inline src/main/java/twilightforest/util/WorldUtil.java:92]: Doesn't it allocate new lambda function instances due to capturing variables like `minX` if the sorting core is moved for instance by a slime block piston flying machine? Or would rewriting this in an imperative style worsen the performance?

File: src/main/java/twilightforest/events/SortingTreeEvents.java
Comment (KaroUniform) [reply src/main/java/twilightforest/events/SortingTreeEvents.java:1]: That's a reasonable suggestion. `MiscEvents` is a better fit for a single stateless handler. I moved the unload listener there and removed the dedicated event class in 73e506b0f. Thanks!

**Comment by Tamaized** [comment]
Just an FYI, I don't want to merge a change like this until the initial 1:1 parity port is complete so this is going to be placed on the back burner but it is being considered regardless.

Though I personally feel some rework needs to be done to *not* have the field inside the block itself. I would love for it to be a block entity instead but that also has some drawbacks we're not willing to make, for example: allowing pistons to push it around. But that's out of scope for this PR and something we'd revisit later.

**Comment by KaroUniform** [comment]
Thanks for the detailed feedback and for clarifying the project's priorities. That makes perfect sense.

If there is any part of the 1:1 parity port where an additional contribution would be helpful, I'd be happy to help. Please feel free to point me toward an issue, area, or task you would recommend.

**Comment by a6tumn** [comment]
Reviewing the open porting PRs that target the *latest* branch is of high priority at the moment. If you are knowledgeable about any of the topics they involve, perhaps you would be able to contribute some feedback? If not, I'm sure there will be more opportunities for contribution down the road, after the porting is finished.
